ports-of-call/static_vector.hpp

```diff
@@ -0,0 +1,424 @@
+    clear();
+    for (auto const &element : other) {
+      push_back(element);
+    }
+  }
+
+  // move constructor
+  // ---------------------------------------------------------------------------
+  PORTABLE_FUNCTION constexpr static_vector(static_vector &&other) noexcept {
+    clear();
+    for (auto &&element : other) {
+      push_back(element);
+    }
+  }
+
+  // copy assignment
+  // ----------------------------------------------------------------------------
+  PORTABLE_FUNCTION constexpr static_vector &operator=(static_vector const &other) {
+    clear();
+    for (auto const &element : other) {
+      push_back(element);
+    }
+    return *this;
+  }
+
+  // move assignment
+  // ----------------------------------------------------------------------------
+  PORTABLE_FUNCTION constexpr static_vector &operator=(static_vector &&other) noexcept {
+    clear();
+    for (auto &&element : other) {
+      push_back(element);
+    }
+    return *this;
+  }
+
+  // destructor
+  // ---------------------------------------------------------------------------------
+  PORTABLE_FUNCTION ~static_vector() { clear(); }
+
+  // operator[]
+  // ---------------------------------------------------------------------------------
+  PORTABLE_FUNCTION constexpr reference operator[](size_type const i) & { return ref(i); }
+
+  PORTABLE_FUNCTION constexpr const_reference operator[](size_type const i) const & {
+    return ref(i);
+  }
+
+  PORTABLE_FUNCTION constexpr Element &&operator[](size_type const i) && {
+    return std::move(ref(i));
+  }
+
+  PORTABLE_FUNCTION constexpr Element const &&operator[](size_type const i) const && {
+    return std::move(ref(i));
+  }
+
+  // front
+  // --------------------------------------------------------------------------------------
+  PORTABLE_FUNCTION constexpr reference front() & { return ref(); }
+
+  PORTABLE_FUNCTION constexpr const_reference front() const & { return ref(); }
+
+  PORTABLE_FUNCTION constexpr Element &&front() && { return ref(); }
+
+  PORTABLE_FUNCTION constexpr Element const &&front() const && { return ref(); }
+
+  // back
+  // ---------------------------------------------------------------------------------------
+  PORTABLE_FUNCTION constexpr reference back() & { return ref(count_ - 1); }
+
+  PORTABLE_FUNCTION constexpr const_reference back() const & { return ref(count_ - 1); }
+
+  PORTABLE_FUNCTION constexpr Element &&back() && { return ref(count_ - 1); }
+
+  PORTABLE_FUNCTION constexpr Element const &&back() const && { return ref(count_ - 1); }
+
+  // data
+  // ---------------------------------------------------------------------------------------
+  PORTABLE_FUNCTION constexpr pointer data() { return ptr(); }
+
+  PORTABLE_FUNCTION constexpr const_pointer data() const { return ptr(); }
+
+  // begin
+  // --------------------------------------------------------------------------------------
+  PORTABLE_FUNCTION constexpr iterator begin() { return {ptr()}; }
```

Comment:
what is the purpose of the braces @BrendanKKrueger ?

Reply:
Probably cut-and-paste from an earlier implementation of static vector that someone else wrote.  Presumably it was intended to indicate that the pointer is an argument to the constructor of the iterator?  But it doesn't appear to be necessary, as I deleted the braces on line 321 and the code seemed to build and run fine.  Would you like me to strip all those extra braces?

Reply:
PS: I say "that someone else wrote" because `git blame` in Singe claims I wrote the file.  I don't recognize a lot of this code, so either my memory sucks (always possible with me) or `git blame` is confused because I did a reformat or similar.  I can't answer for every decision in the code, but I'm happy to iterate with you to adapt the code to something that "fits" with Ports-of-Call.

Reply:
My preference is to remove the braces if unneeded, but I can be convinced otherwise if there are good arguments for it.

Reply:
Deleting the braces seemed to have no effect, so I'm not aware of any arguments to keep them.